From 0e8cab9c39c559c9616b42619492efaaf90c968b Mon Sep 17 00:00:00 2001
From: stephen hemminger <stephen@networkplumber.org>
Date: Wed, 19 Jul 2017 11:53:19 -0700
Subject: [PATCH 37/98] netvsc: add rtnl annotations in rndis

The rndis functions are used when changing device state.
Therefore the references from network device to internal state
are protected by RTNL mutex.

Signed-off-by: Stephen Hemminger <sthemmin@microsoft.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/hyperv/hyperv_net.h   |    6 ------
 drivers/net/hyperv/netvsc.c       |    6 ++++--
 drivers/net/hyperv/netvsc_drv.c   |    1 +
 drivers/net/hyperv/rndis_filter.c |   12 ++++++++++--
 4 files changed, 15 insertions(+), 10 deletions(-)

--- a/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:20:16.106993729 -0800
+++ b/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:20:16.102993782 -0800
@@ -788,12 +788,6 @@ struct netvsc_device {
 	struct rcu_head rcu;
 };
 
-static inline struct netvsc_device *
-net_device_to_netvsc_device(struct net_device *ndev)
-{
-	return ((struct net_device_context *)netdev_priv(ndev))->nvdev;
-}
-
 /* NdisInitialize message */
 struct rndis_initialize_request {
 	u32 req_id;
--- a/drivers/net/hyperv/netvsc.c	2017-11-17 10:20:16.106993729 -0800
+++ b/drivers/net/hyperv/netvsc.c	2017-11-17 10:20:16.102993782 -0800
@@ -108,7 +108,8 @@ static void netvsc_destroy_buf(struct hv
 {
 	struct nvsp_message *revoke_packet;
 	struct net_device *ndev = hv_get_drvdata(device);
-	struct netvsc_device *net_device = net_device_to_netvsc_device(ndev);
+	struct net_device_context *ndc = netdev_priv(ndev);
+	struct netvsc_device *net_device = rtnl_dereference(ndc->nvdev);
 	int ret;
 
 	/*
@@ -831,7 +832,8 @@ int netvsc_send(struct net_device_contex
 		struct hv_page_buffer **pb,
 		struct sk_buff *skb)
 {
-	struct netvsc_device *net_device = rcu_dereference(ndev_ctx->nvdev);
+	struct netvsc_device *net_device
+		= rcu_dereference_rtnl(ndev_ctx->nvdev);
 	struct hv_device *device = ndev_ctx->device_ctx;
 	int ret = 0;
 	struct netvsc_channel *nvchan;
--- a/drivers/net/hyperv/netvsc_drv.c	2017-11-17 10:20:16.106993729 -0800
+++ b/drivers/net/hyperv/netvsc_drv.c	2017-11-17 10:20:16.102993782 -0800
@@ -1576,6 +1576,7 @@ static int netvsc_probe(struct hv_device
 
 	netif_set_real_num_tx_queues(net, nvdev->num_chn);
 	netif_set_real_num_rx_queues(net, nvdev->num_chn);
+	rtnl_unlock();
 
 	netdev_lockdep_set_classes(net);
 
--- a/drivers/net/hyperv/rndis_filter.c	2017-11-17 10:20:16.106993729 -0800
+++ b/drivers/net/hyperv/rndis_filter.c	2017-11-17 10:20:16.102993782 -0800
@@ -84,6 +84,14 @@ static struct rndis_device *get_rndis_de
 	return device;
 }
 
+static struct netvsc_device *
+net_device_to_netvsc_device(struct net_device *ndev)
+{
+	struct net_device_context *net_device_ctx = netdev_priv(ndev);
+
+	return rtnl_dereference(net_device_ctx->nvdev);
+}
+
 static struct rndis_request *get_rndis_request(struct rndis_device *dev,
 					     u32 msg_type,
 					     u32 msg_len)
@@ -472,7 +480,7 @@ static int rndis_filter_query_device(str
 
 	if (oid == OID_TCP_OFFLOAD_HARDWARE_CAPABILITIES) {
 		struct net_device_context *ndevctx = netdev_priv(dev->ndev);
-		struct netvsc_device *nvdev = ndevctx->nvdev;
+		struct netvsc_device *nvdev = rtnl_dereference(ndevctx->nvdev);
 		struct ndis_offload *hwcaps;
 		u32 nvsp_version = nvdev->nvsp_version;
 		u8 ndis_rev;
@@ -944,7 +952,7 @@ static void rndis_filter_halt_device(str
 	struct rndis_request *request;
 	struct rndis_halt_request *halt;
 	struct net_device_context *net_device_ctx = netdev_priv(dev->ndev);
-	struct netvsc_device *nvdev = net_device_ctx->nvdev;
+	struct netvsc_device *nvdev = rtnl_dereference(net_device_ctx->nvdev);
 
 	/* Attempt to do a rndis device halt */
 	request = get_rndis_request(dev, RNDIS_MSG_HALT,
